overlord/configstate/configcore/tmp.go

```diff
@@ -0,0 +1,145 @@
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package configcore
+
+import (
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"os/exec"
+	"path/filepath"
+
+	"github.com/snapcore/snapd/dirs"
+	"github.com/snapcore/snapd/gadget/quantity"
+	"github.com/snapcore/snapd/osutil"
+	"github.com/snapcore/snapd/overlord/configstate/config"
+	"github.com/snapcore/snapd/sysconfig"
+)
+
+const (
+	mntStaticOptions         = "mode=1777,strictatime,nosuid,nodev"
+	tmpfsMountPoint          = "/tmp"
+	tmpMntServOverrideSubDir = "tmp.mount.d"
+	tmpMntServOverrideFile   = "override.conf"
+)
+
+func init() {
+	// add supported configuration of this module
+	supportedConfigurations["core.tmp.size"] = true
+}
+
+func validTmpfsSize(sizeStr string) error {
+	if sizeStr == "" {
+		return nil
+	}
+
+	size, err := quantity.ParseSize(sizeStr)
+	if err != nil {
+		return err
+	}
+
+	// Do not allow less than 16mb
+	// 0 is special and means unlimited
+	if size > 0 && size < 16*quantity.SizeMiB {
+		return fmt.Errorf("size is less than 16Mb")
+	}
+
+	return nil
+}
+
+func validateTmpfsSettings(tr config.ConfGetter) error {
+	tmpfsSz, err := coreCfg(tr, "tmp.size")
+	if err != nil {
+		return err
+	}
+
+	return validTmpfsSize(tmpfsSz)
+}
+
+func handleTmpfsConfiguration(_ sysconfig.Device, tr config.ConfGetter, opts *fsOnlyContext) error {
+	tmpfsSz, err := coreCfg(tr, "tmp.size")
+	if err != nil {
+		return err
+	}
+
+	// Create override configuration file for tmp.mount service
+
+	// Create /etc/systemd/system/tmp.mount.d if needed
+	var overrDir string
+	if opts == nil {
+		// runtime system
+		overrDir = dirs.SnapServicesDir
+	} else {
+		overrDir = dirs.SnapServicesDirUnder(opts.RootDir)
+	}
+	overrDir = filepath.Join(overrDir, tmpMntServOverrideSubDir)
+
+	// Write service config override if needed
+	options := mntStaticOptions
+	dirContent := make(map[string]osutil.FileState, 1)
+	cfgFilePath := filepath.Join(overrDir, tmpMntServOverrideFile)
+	modify := true
+	if tmpfsSz != "" {
+		if err := os.MkdirAll(overrDir, 0755); err != nil {
+			return err
+		}
+		options = fmt.Sprintf("%s,size=%s", options, tmpfsSz)
+		content := fmt.Sprintf("[Mount]\nOptions=%s\n", options)
+		dirContent[tmpMntServOverrideFile] = &osutil.MemoryFileState{
+			Content: []byte(content),
+			Mode:    0644,
+		}
+		oldContent, err := ioutil.ReadFile(cfgFilePath)
+		if err == nil && content == string(oldContent) {
+			modify = false
+		}
+	} else {
+		// Use default tmpfs size if empty setting (50%, see tmpfs(5))
```

Comment:
it would be sort of nice when the setting is empty if we returned the value as 50%. For swap, the default is 0 bytes of swap so it is logical that the value we return is `"0"` or `""`, but the default for tmpfs is not 0 bytes, so the default value returned should be 50% I think

Reply:
hmm although just checking this on a local system, when swap is not set we get an error checking the value:

```
root@ubuntu:/# snap get system swap.size
error: snap "core" has no "swap" configuration option
```

which I guess is consistent with other system settings but is kinda unexpected

Reply:
Right, that is also what I get for an unset `tmp.size`. However, I get an empty line if I do a get after `snap set system tmp.size=`. But you cannot distinguish between the empty and unset case inside `tmp.go`, I think. Also, I think that to do what you suggest we would need a way for the settings code to inform of the default value for each setting.